src/main/java/com/redhat/red/build/finder/BuildFinder.java

```diff
@@ -117,31 +138,220 @@ public BuildFinder(ClientSession session, BuildConfig config) {
+            KojiLocalArchive existingArchive = matchingArchive.get();
+            filenames = existingArchive.getFiles();
+
+            LOGGER.debug("Adding not-found checksum {} to existing archive {}", existingArchive.getArchive().getArchiveId());
+
+            filenames.addAll(files);
+        } else {
+            filenames = new ArrayList<>(files);
+
+            KojiArchiveInfo tmpArchive = new KojiArchiveInfo();
+
+            tmpArchive.setBuildId(0);
+            tmpArchive.setFilename("not found");
+            tmpArchive.setChecksum(checksum);
+            tmpArchive.setChecksumType(config.getChecksumType());
+
+            tmpArchive.setArchiveId(-1 * (buildZero.getArchives().size() + 1));
+
+            LOGGER.debug("Adding not-found checksum {} to new archive {}", tmpArchive.getArchiveId());
+
+            buildZero.getArchives().add(new KojiLocalArchive(tmpArchive, filenames));
+        }
+
+        filenames.sort(Comparator.naturalOrder());
+    }
+
+    private void addArchiveToBuild(KojiBuild build, KojiArchiveInfo archive, Collection<String> files) {
+        List<String> filenames = new ArrayList<>(files);
+
+        LOGGER.debug("Adding all filenames with archive id {} with files {}", archive.getArchiveId(), filenames);
+
+        if (build.getArchives() == null) {
+            build.setArchives(new ArrayList<>());
+        }
+
+        Optional<KojiLocalArchive> matchingArchive = build.getArchives().stream().filter(a -> a.getArchive().getArchiveId().equals(archive.getArchiveId())).findFirst();
+
+        if (matchingArchive.isPresent()) {
+            KojiLocalArchive existingArchive = matchingArchive.get();
+
+            if (existingArchive.getFiles() == null) {
+                LOGGER.debug("Setting up new filenames with archive id {} with files {}", archive.getArchiveId(), filenames);
+                existingArchive.setFiles(filenames);
+            } else {
+                LOGGER.debug("Adding to existing filenames for archive id {}", archive.getArchiveId());
+                existingArchive.getFiles().addAll(filenames);
+            }
+
+            existingArchive.getFiles().sort(Comparator.naturalOrder());
+
+            LOGGER.debug("Added to existing archive id {} to build id {} with {} archives", archive.getArchiveId(), archive.getBuildId(), build.getArchives().size());
+        } else {
+             build.getArchives().add(new KojiLocalArchive(archive, filenames));
 
-        int checked = 0;
-        int numChecksums = checksumTable.size();
-        int hits = 0;
+            LOGGER.debug("Added new archive id {} to build id {} with {} archives", archive.getArchiveId(), archive.getBuildId(), build.getArchives().size());
+        }
+
+        build.getArchives().sort((KojiLocalArchive a1, KojiLocalArchive a2) -> a1.getArchive().getFilename().compareTo(a2.getArchive().getFilename()));
+    }
+
+    private KojiBuild findBestBuild(List<KojiBuild> foundBuilds, List<KojiArchiveInfo> archives) {
```

Comment:
This is the method doing most of the new work here. The rest of the changes are mostly related to reorganization.

Duplicate builds (that is, builds containing the same archives) are being stored still. Nothing is done with these. Future work may be to consolidate these builds.

Reply:
Ok. Can we document the order selection for posterity?

Reply:
You mean as a Java comment? I thought about imposing a total ordering on builds that does not rely on the id, maybe as a standalone comparator, but the order intended here is:

COMPLETE !IMPORT TAGGED
COMPLETE IMPORT TAGGED
COMPLETE !IMPORT
COMPLETE IMPORT
ID

Note that builds are already returned by Koji ordered by id. Incomplete builds are currently considered by id only. Imported tagged builds are considered better than non-imported untagged builds under the presumption that an untagged build will be deleted.

Reply:
+1 just to show the intention of what is considered 'better'.